### PR TITLE
Refactor CMake and introduce xtensor

### PR DESCRIPTION
**Pull request by garth-wells**
The PR has a lot in it - it started as introducing xtensor, but fixed quite a few other bugs along the way. These bugs would have made it hard to make the CMake changes off the master branch.

### CODE REVIEW

File: .github/workflows/dolfin-tests.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # This workflow will install Basix, FFCx, DOLFINx and run the DOLFINx unit tests.
 
-name: Basix CI
+name: Basix/DOLFINX CI
```
Comment (mscroggs) [inline .github/workflows/dolfin-tests.yml:3]: Elsewhere (conference, paper), we've been writing DOLFINx with lowercase x

File: cpp/basix/finite-element.cpp [hunk 2]
```diff
@@ -84,7 +86,17 @@ Eigen::MatrixXd basix::compute_expansion_coefficients(
...
+  if (x.cols() > 1)
+    shape.push_back(x.cols());
+  auto _x = xt::adapt<xt::layout_type::column_major>(x.data(), x.size(),
+                                                     xt::no_ownership(), shape);
+  const xt::xtensor<double, 3> PP = polyset::tabulate(celltype, degree, 0, _x);
+  Eigen::MatrixXd P(PP.shape()[1], PP.shape()[2]);
+  for (std::size_t i = 0; i < PP.shape()[1]; ++i)
+    for (std::size_t j = 0; j < PP.shape()[2]; ++j)
+      P(i, j) = PP(0, i, j);
+  // const Eigen::MatrixXd P = polyset::tabulate(celltype, degree, 0, x)[0];
```
Comment (mscroggs) [inline cpp/basix/finite-element.cpp:99]: Should this commented line be removed?

File: cpp/basix/serendipity.cpp [hunk 3]
```diff
@@ -26,10 +26,13 @@ Eigen::MatrixXd make_serendipity_space_2d(const int degree)
   // Evaluate the expansion polynomials at the quadrature points
   auto [Qpts, Qwts] = quadrature::make_quadrature(
       "default", cell::type::quadrilateral, 2 * degree);
+  std::cout << "S0 call" << std::endl;
```
Comment (mscroggs) [inline cpp/basix/serendipity.cpp:29]: Should these couts still be here?

File: cpp/basix/regge.cpp [hunk 4]
```diff
@@ -73,34 +74,42 @@ create_regge_interpolation(cell::type celltype, int degree)
...
-          {
-            points(point_n + j, p)
-                += (entity_geom(k + 1, p) - entity_geom(0, p)) * lattice(j, k);
-          }
+          // for (std::size_t p = 0; p < entity_geom.shape()[1]; ++p)
+          // {
+          //   points(point_n + j, p)
+          //       += (entity_geom(k + 1, p) - entity_geom(0, p)) * lattice(j,
+          //       k);
+          //
```
Comment (mscroggs) [inline cpp/basix/regge.cpp:96]: Delete this old commented code?

File: cpp/basix/regge.cpp [hunk 5]
```diff
@@ -73,34 +74,42 @@ create_regge_interpolation(cell::type celltype, int degree)
...
+          xt::row(points, point_n + j)
+              += (xt::row(entity_geom, k + 1) - xt::row(entity_geom, 0))
+                 * lattice(j, k);
           // points.row(point_n + j)
           //     += (entity_geom.row(k + 1) - entity_geom.row(0)) * lattice(j,
           //     k);
         }
```
Comment (mscroggs) [inline cpp/basix/regge.cpp:103]: More commented code

**Review by chrisrichardson — APPROVED** [review]

File: test/test_orthonormal_basis.py [hunk 6]
```diff
@@ -25,8 +25,8 @@ def test_quad(order):
         for j in range(ndofs):
             mat[i, j] = sum(basis[:, i] * basis[:, j] * Qwts)
 
-    np.set_printoptions(suppress=True)
-    print(mat, np.eye(mat.shape[0]))
+    # np.set_printoptions(suppress=True)
+    # print(mat, np.eye(mat.shape[0]))
```
Comment (mscroggs) [inline test/test_orthonormal_basis.py:29]: Should the comments in these tests be removed?

File: test/test_orthonormal_basis.py [hunk 7]
```diff
@@ -98,26 +98,30 @@ def test_prism(order):
...
-@pytest.mark.parametrize("order", [0, 1, 2, 3, 4])
+@pytest.mark.parametrize("cell_type", [
+    basix.CellType.interval,
+    basix.CellType.triangle,
+    basix.CellType.tetrahedron
+])
+# @pytest.mark.parametrize("order", [0, 1, 2, 3, 4])
+@pytest.mark.parametrize("order", [1])
 def test_cell(cell_type, order):
     Qpts, Qwts = basix.make_quadrature("default", cell_type, 2*order + 1)
```
Comment (mscroggs) [inline test/test_orthonormal_basis.py:114]: Missing space around `*`. Shouldn't the flake8 checks pick this up? Could be a sign that flake8 isn't running everywhere?

**Review by mscroggs — APPROVED** [review]

**Comment by mscroggs** [comment]
setup-py and cmake need removing from requirements, as they no longer exist. I'll do that now

**Pull request merged**